### ISSUE TITLE
Inaccurate visit/playout limiting when using multiple threads

### ISSUE THREAD

**Issue by AncalagonX** [gcp/leela-zero#2064]
I'll report two related minor problems in this thread. At first I thought they were the exactly the same issue, but there seems to be a larger overrun in the first case than in the second case. They may not be directly related.

(1) It seems that `-v 800 -t 4` or `-p 800 -t 4` will rarely, if ever, terminate exactly at the precise desired number of playouts or visits. This can readily be seen at the end of the turn when unmodified vanilla LZ outputs its stats into the GTP log. Quick testing will find that the number of **playouts** in a random mid-game position can exceed (playouts + threads)!

I imagine counting for accurate visits and then implementing an accurate cut-off is like herding cats to some degree when dealing with multiple (unsynchronized?) threads making their own visits, but this slight inaccuracy is probably something that should be hammered out at some point.

**Comment by AncalagonX** [comment]
When testing (1) using unmodified LZ with the settings `-v 800 -t 4 --noponder` on a random mid-game move that LZ didn't expect, the search *should* terminate at 800 playouts, or at most due to the above inaccuracy at a max of 800 + 4 = 804 playouts. However, it can sometimes end up with **805 playouts** or **806 playouts** which exceeds expectations beyond what I can explain.

(2) Additionally, something related affects me when I am heavily modifying UCTNode.cpp for my own purposes. When I add conditionals like `if (child.get_visits() >= 400) { then go_do_something }` into the for loop in `uct_select_node`, I experience the same difficulty in getting an accurate visit count while using multiple threads. However, it manifests slightly differently than (1) above.

What's especially unusual is that this behavior contrasts with the way that inaccurate visit accounting manifests when modifying LZ as in (2), where I add in conditionals based on visit counts into `uct_select_node`. In those cases, I can end up with nodes which received 400, 401, 402, or 403 visits when using 4 threads, averaging an overrun of (threads/2), but *never* an "excess of excess" at 405 or 406 visits like seen at the end-of-turn GTP stats output for playouts in (1) above.

Note: Though I used `-v` in my examples and then at times spoke in terms of "playouts" and other times in terms of "visits", there shouldn't be any accidental typos in word choice above. I'm not using the terms interchangeably—playouts are playouts, and visits are visits.

**Comment by AncalagonX** [comment]
@sethtroisi mentioned on Discord re: (1) for inaccurate visit/playout limits in unmodified LZ:

> I don't think the "problem" is synchronizating the counter but the asyncronous nature of network.cpp. 

>I suspect "fixing" this would be as simple as changing `m_root->get_visits() >= m_maxvisits` to `m_root->get_visits() +m_root->virtual_losses >= m_maxvisits`

But I'm not familiar with the code underpinning this issue.

**Comment by alreadydone** [comment]
2. For each thread, it first checks the visits (what you implemented) in uct_select_child (your typo here?), and then after a simulation is over, updates the visits, so each thread goes check->update->check->update->... The visits are incremented by one at a time, so at some moment it must be 400 exactly. In order to reach 404, the visits must be updated four times in a row after this moment, so there are five consecutive updates without checking in the meantime, and it's not hard to see there must be at least five threads running in parallel.

1. The difference with 2 is that not all threads check the visits; only the main thread does, and there's no limit on the amount of things the other threads can do between two checks of the main thread. (The main thread also outputs playouts, winrate, and principal variation in addition to playing simulations.)

3. I think you want `UCTNode::VIRTUAL_LOSS_COUNT * m_root->get_visits() + m_root->m_virtual_loss >= UCTNode::VIRTUAL_LOSS_COUNT * m_maxvisits`, but even then you are not guaranteed to get equality, since m_virtual_loss and m_visits aren't updated at the same time (atomically or protected by lock); you may have started a simulation but m_virtual_loss hasn't been incremented, or you may have updated the visits but the virtual loss hasn't been undone. (BTW I don't see anything asynchronous in Network.cpp in vanilla LZ; my 1t-batch and fastexit branches make it asynchronous though.)

I wonder what's the benefit of making things terminate at the exact count; I think you need to be clever to achieve that without reducing efficiency.
without reducing efficiency

**Comment by sethtroisi** [comment]
to expand on my discord comment. I think the issue is this

virtual loss = vloss = current playouts being evaluated by the network 
with 100 playout limit

T1: 99 playouts I can start another play_simulation
T2: 99 playouts + 1 vloss I can start another play_simulation
T3: 99 playouts + 2 vloss I can start another play_simulation
T1: finishes simulation, decrements virtual loss, increments playouts | 100 playouts + 2 vloss (T2, T3) I shouldn't do another simulation, loop breaks
T2: finishes simulation, decrements virtual loss, increments playouts | 101 playous + 1 vloss (T3)
T3: finishes simulation, decrements virtual loss, increments playouts | 102 playous + 0 vloss

Relevant code links
[loop](https://github.com/gcp/leela-zero/blob/master/src/UCTSearch.cpp#L745)
[m_root->get_visits() < m_maxvisits](https://github.com/gcp/leela-zero/blob/master/src/UCTSearch.cpp#L668)

**Comment by ihavnoid** [comment]
@AncalagonX - Understood that it is inaccurate, but can you elaborate what value does it may bring if we make it really accurate?

**Comment by ihavnoid** [comment]
Looking at the source code, the reason that it can go beyond 804 visits is because the only thread that can terminate the playouts loop is the main thread.  If for any reason the main thread is slower than any other thread (e.g., Every other thread ended up hitting on the NNCache) the playouts can go up as much as it wants.

**Comment by gcp** [comment]
This was/is pretty much by design. 

>I wonder what's the benefit of making things terminate at the exact count; I think you need to be clever to achieve that without reducing efficiency.

Exactly: I wasn't clever enough.

If you are at a low enough playout count that the excess matters, then I doubt you want the influence of the parallel search in the first place. Making this more accurate without efficiency impact is fine, but so is WONTFIX.
